compiler/kernel/bufferfilter.go

```diff
@@ -73,41 +80,41 @@ func CompileBufferFilter(e dag.Expr) (*expr.BufferFilter, error) {
 	}
 }
 
-func isFieldEqualOrIn(e *dag.BinaryExpr) (*astzed.Primitive, string) {
+// XXX isFieldEqualOrIn should work for any paths not just top-level fields.
+// See issue #3412
+
+func isFieldEqualOrIn(zctx *zed.Context, e *dag.BinaryExpr) (*zed.Value, error) {
 	if dag.IsTopLevelField(e.LHS) && e.Op == "=" {
-		if literal, ok := e.RHS.(*astzed.Primitive); ok {
-			return literal, "="
+		if literal, ok := e.RHS.(*dag.Literal); ok {
+			val, err := zson.ParseValue(zctx, literal.Value)
+			if err != nil {
+				return nil, err
+			}
+			return &val, nil
 		}
 	} else if dag.IsTopLevelField(e.RHS) && e.Op == "in" {
-		if literal, ok := e.LHS.(*astzed.Primitive); ok && literal.Type != "net" {
-			return literal, "in"
+		if literal, ok := e.LHS.(*dag.Literal); ok {
+			val, err := zson.ParseValue(zctx, literal.Value)
+			if err != nil {
+				return nil, err
+			}
+			if val.Type == zed.TypeNet {
+				return nil, err
+			}
+			return &val, nil
 		}
 	}
-	return nil, ""
+	return nil, nil
 }
 
-func newBufferFilterForLiteral(l astzed.Primitive) (*expr.BufferFilter, error) {
-	switch l.Type {
-	case "bool", "byte", "int16", "uint16", "int32", "uint32", "int64", "uint64",
-		"float32", "float64", "time", "duration":
-		// These are all comparable, so they can require up to three
+func newBufferFilterForLiteral(val *zed.Value) (*expr.BufferFilter, error) {
+	if id := val.Type.ID(); zed.IsNumber(id) || id == zed.IDNull {
+		// All numbers are all comparable, so they can require up to three
```

Comment:
This typo was in my suggestion. Sorry about that.
```suggestion
		// All numbers are comparable, so they can require up to three
```